cli/internal/core/engine.go

```diff
@@ -385,11 +410,76 @@ func (e *Engine) ValidatePersistentDependencies(graph *graph.CompleteGraph) erro
+func (e *Engine) getTaskDefinitionChain(rootPipeline *fs.Pipeline, pkg *fs.PackageJSON, taskID string, taskName string) ([]fs.BookkeepingTaskDefinition, error) {
+	// Start a list of TaskDefinitions we've found for this TaskID
+	taskDefinitions := []fs.BookkeepingTaskDefinition{}
+
+	// Look for the taskDefinition in the root pipeline.
+	// We'll wait to throw errors until the end
+	if rootTaskDefinition, err := rootPipeline.GetTask(taskID, taskName); err == nil {
+		taskDefinitions = append(taskDefinitions, *rootTaskDefinition)
+	}
+
+	// If the taskID is a root task (e.g. //#build), we don't need to look
+	// for a workspace task, since these can only be defined in the root turbo.json.
+	taskIDPackage, _ := util.GetPackageTaskFromId(taskID)
+	if taskIDPackage != util.RootPkgName && taskIDPackage != ROOT_NODE_NAME {
+		// Look up task definition in turbo.json in the workspace directory
+		workspaceConfigPath := pkg.Dir.Join("turbo.json").RestoreAnchor(e.completeGraph.RepoRoot)
+
+		// If there is an error, we can ignore it, since turbo.json config is not required in the workspace.
+		if workspaceTurboJSON, err := fs.ReadTurboConfig(workspaceConfigPath); err == nil {
+			// TODO(mehulkar): Enable extending from more than one workspace.
+			if len(workspaceTurboJSON.Extends) > 1 {
+				return nil, fmt.Errorf(
+					"You can only extend from the root workspace. \"%s\" extends from %v",
+					pkg.Name,
+					workspaceTurboJSON.Extends,
+				)
+			}
+
+			// TODO(mehulkar):
+			// 		Pipeline.GetTask allows searching with a taskID (e.g. `package#task`).
+			// 		But we do not want to allow this, except if we're in the root workspace.
+			workspaceDefinition, err := workspaceTurboJSON.Pipeline.GetTask(taskID, taskName)
+
+			if err == nil {
+				taskDefinitions = append(taskDefinitions, *workspaceDefinition)
+			}
+
+			// If there is no Extends key, we are either in a workspace that didn't
+			// extend from anything, or in a single package repo, where the workspace is the root.
+			if len(workspaceTurboJSON.Extends) == 0 {
+				// For the single package case, we can just return the workspaceDefinition
+				// (It should be the same as the rootTaskDefinition)
+				if pkg.PackageJSONPath == "package.json" {
+					return []fs.BookkeepingTaskDefinition{*workspaceDefinition}, nil
+				}
+
+				// For turbo.jsons in workspaces that don't have an extends key, throw an error
+				// We don't support this right now.
+				return nil, fmt.Errorf("No \"extends\" key found in %s", workspaceConfigPath)
+			}
+
+			// TODO(mehulkar): Enable extending from non-root workspace.
+			if workspaceTurboJSON.Extends[0] != util.RootPkgName {
+				return nil, fmt.Errorf(
+					"You can only extend from the root workspace. \"%s\" extends from %v",
+					pkg.Name,
+					workspaceTurboJSON.Extends,
+				)
+			}
+		}
+	}
+
+	if len(taskDefinitions) == 0 {
+		return nil, fmt.Errorf("Could not find \"%s\" in root turbo.json or \"%s\" workspace", taskID, pkg.Dir)
```

Comment:
We might want to special case missing a root task definition here, since there's no workspace for it to be in.

Reply:
Do you mean for single package repos? Otherwise, I think the error message says: "you can define it in `./turbo.json` or `./apps/web/turbo.json` and both are ok", which seems like a good teaching moment.